DTS-HD parsing

Nadahar opened the pull request:
@Sami32 I've tried to resurrect https://github.com/UniversalMediaServer/UniversalMediaServer/pull/1274 here.

The reason https://github.com/UniversalMediaServer/UniversalMediaServer/pull/1274 stalled was because you had made a whole lot of changes that had nothing to do with DTS-HD and which created problems and disagreements. I've tried to separate out just the changes in one commit and the formatting changes in another. The rest of the non-related changes I've removed from this PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/digitalmediaserver/digitalmediaserver/43)
<!-- Reviewable:end -->

Sami32 commented:
LOL...
It was very simple at first, before you ask for changes and changes...that created tests issues that i need to try to solve...

Sami32 commented on src/main/external-resources/renderers/DefaultRenderer.conf:
```diff
@@ -582,8 +582,8 @@ CharMap =
-#    dts      Digital Theater Systems                       audio/vnd.dts
-#    dtshd    Digital Theater Systems (HD)                  audio/vnd.dts.hd
+#    dts      Dedicated To Sound                            audio/vnd.dts                   Including DTS Coherent Acoustics (DCA)
+#    dtshd    Dedicated To Sound High Definition            audio/vnd.dts.hd                Including X, MA, HRA and Core
```
I should have remove Core

Sami32 commented on src/main/external-resources/renderers/DefaultRenderer.conf:
```diff
@@ -582,8 +582,8 @@ CharMap =
 #    dsf      DSD Storage Facility                          audio/x-dsf
-#    dts      Digital Theater Systems                       audio/vnd.dts
-#    dtshd    Digital Theater Systems (HD)                  audio/vnd.dts.hd
+#    dts      Dedicated To Sound                            audio/vnd.dts                   Including DTS Coherent Acoustics (DCA)
```
I should have writed "Aka DTS Coherent..."

Sami32 commented on src/main/java/net/pms/dlna/DLNAMediaInfo.java:
```diff
@@ -1770,7 +1772,16 @@ public boolean isVideoWithinH264LevelLimits(InputFile f, RendererConfiguration m
+				codecA.startsWith("dts") ||
+				codecA.startsWith("a_dts") ||
+				codecA.equals("dca") ||
+				codecA.contains("flac")
```
I should have writed:
```
"flac".equals(codecA) ||
"19d".equals(codecA)
```

Nadahar commented:
@Sami32 It wasn't "very simple" at first, because you had made many changes creating new fields in ```DLNAMediaInfo``` and ```AudioProperties``` without implementing these all the way (database etc) that from what I understand was only there to get better parse logging.

Since this is now handled by the parselogger, I simply removed this, and that cleared up a lot.

Sami32 commented on src/main/java/net/pms/dlna/DLNAMediaInfo.java:
```diff
@@ -1770,7 +1772,16 @@ public boolean isVideoWithinH264LevelLimits(InputFile f, RendererConfiguration m
+			) &&
+			!codecA.contains("pcm_u8") &&
+			!codecA.contains("pcm_s8");
 	}
```
I should have said that not only FFmpeg is used for parsing so it will not work for MediaInfo ;P

Sami32 commented:
I'll have to look at my original code later, because it seem that you have broken stuff.

Sami32 commented:
Oh i see, you removed the bitrate and delay fixes. Well it is too old for me now for understanding all i've done.

Nadahar commented:
@Sami32 I didn't see any bitrate and delay fixes. When you say fixes, to you mean that you had changed the parsing, or that there were some other bugs?

It's easy to see the "original" PR, I kept the "raw" cherry-picked version here: https://github.com/DigitalMediaServer/DigitalMediaServer/commits/DTS-HD_parsing_and_bitrate_mess

Just tell me what parts you're missing and I'll "transfer" them.

Sami32 commented:
Audio delay has been broken, not working since years so...
Default values are missing as well, but it is already too long for me to understand all this code again, so will be better than nothing for now.

Nadahar commented:
@Sami32 I've taken the parsers from the old PR, but put them in ```LibMediaInfoParser``` where I think they belong. They are slightly modified, and I created at least one new. I think they should be fine, but I need your help to decide for which fields we should use the *first* value and which we should use the *last* when there are multiple values.

I've also moved all fields the ```DLNAMediaAudio``` so that ```AudioProperties``` shouldn't be used and can be removed later. I've implemented default values as you wanted in ```DLNAMediaAudio``` although I'm against that in principle. When you don't know something, you shouldn't pretent to know by returning a default - instead the specific code that use it should use a default value if needed. I handled this by making a ```raw``` version and ```isXxxUnknown()``` methods of those fields with defaults. 

Storing in the database is an excellent example of why using defaults is wrong. The database is now using the ```raw``` methods so it doesn't store a default. If a default was stored and this record was later retrieved, there would be know way to tell if the value was real or just a default.

I hope this will please you, and it would be nice if you could do some tests on relevant files.

Sami32 commented:
I agree that in a perfect world we shouldn't need to implement default values, but we need to workaround parsing issues, like with MediaInfo, to respect the minimum DLNA specifications to make DMS able to stream without failing.
As we depends on external libraries, it seemed to me the more proper way to fix that.

That's said, i'm open to any better suggestion ;)

Nadahar commented:
My point isn't that I'm against default values, I'm against storing them as actual data. In my opinion only actual data should be stored so that the code that uses the data can know if they are known or unknown. This code then needs the decide if a default should be used if the data is missing, if the information can simply be dropped or if an error should occur.

When storing default values *as if they were real, parsed values* there's not way for the code that use this to know the difference between real values and (fake) default values.

Sami32 commented:
As long as the default values are used for the DIDL strings when need, it is fine for me.

Sami32 commented on src/main/java/net/pms/dlna/DLNAMediaInfo.java:
```diff
@@ -1195,19 +1195,23 @@ public void parseFFmpegInfo(List<String> lines, String input) {
+								}
 							} else if (token.equals("mono")) {
-								audio.getAudioProperties().setNumberOfChannels(1);
+								audio.setNumberOfChannels(1);
```
Not related to this PR, but that make me wondering if a better way to handle the channels can be found...
Because many are omit, like 3 channels, 7 or 9 channels and so on.

Nadahar replied:
That would depends on the FFmpeg output. I don't know if it's possible to get a number instead of a text - if it's not possible, I guess we just need to handle all the different texts....

Nadahar replied:
@Sami32 As far as I can see, there's no way to get a more meaningful output from FFmpeg. FFProbe on the other hand has possibilities, but that would require yet another binary:
```ffprobe.exe -v quiet -print_format json -show_format -show_streams <file>```

Sami32 replied:
No need, i was referring to FFmpeg and MediaInfo informations delivered:
Channel(s)                               : 3 channels
Channel(s)                               : 8 channels
Channel(s)                               : 4 channels
Channel(s)                               : 10 channels
and so on

Nadahar replied:
Then I don't quite understand what you mean, MediaInfo will simply return the number of channels - which we should be able to parse just fine. Of course we can't tell the difference between 2.1 and 3.0, was that what you meant?

For FFmpeg it's worse since we don't get the number, but a string which we would then have to match every variant of (like "mono", "stereo" etc).

Sami32 replied:
From FFmpeg:
Standard channel layouts:
NAME           DECOMPOSITION
mono           FC
stereo         FL+FR
2.1            FL+FR+LFE
3.0            FL+FR+FC
3.0(back)      FL+FR+BC
4.0            FL+FR+FC+BC
quad           FL+FR+BL+BR
quad(side)     FL+FR+SL+SR
3.1            FL+FR+FC+LFE
5.0            FL+FR+FC+BL+BR
5.0(side)      FL+FR+FC+SL+SR
4.1            FL+FR+FC+LFE+BC
5.1            FL+FR+FC+LFE+BL+BR
5.1(side)      FL+FR+FC+LFE+SL+SR
6.0            FL+FR+FC+BC+SL+SR
6.0(front)     FL+FR+FLC+FRC+SL+SR
hexagonal      FL+FR+FC+BL+BR+BC
6.1            FL+FR+FC+LFE+BC+SL+SR
6.1(back)      FL+FR+FC+LFE+BL+BR+BC
6.1(front)     FL+FR+LFE+FLC+FRC+SL+SR
7.0            FL+FR+FC+BL+BR+SL+SR
7.0(front)     FL+FR+FC+FLC+FRC+SL+SR
7.1            FL+FR+FC+LFE+BL+BR+SL+SR
7.1(wide)      FL+FR+FC+LFE+BL+BR+FLC+FRC
7.1(wide-side) FL+FR+FC+LFE+FLC+FRC+SL+SR

Sami32 replied:
Though i think remember some standard like MPEG-4 audio supporting 15 channels and that FFmpeg have also implemented it, but i don't know how it is displayed. Something like 9, 10 11, 12, 13, 14, 15 i guess

Nadahar replied:
The code you have made this comment on is fro the FFmpeg parsing, inside the parsing for "mono". There is such a parsing, although it will only cover a few of the most common strings. To maintain such a list of values is a bad solution IMO, although it seems the only one available unless we use FFprobe instead. As it is now, most of the above options aren't parsed.

This has nothing to do with MediaInfo that I understand though, and I don't think we should address it in this PR anyway.

Sami32 replied:
https://github.com/FFmpeg/FFmpeg/blob/977fd8841921f0109bee220e1f99dd662c166ae1/libavutil/channel_layout.c

Sami32 replied:
It was only a question, remember.
I'm not in favor to add an other executable.
And as you always tease me saying that MediaInfo is not the only tool used ;PP
Yes, nothing to do in that PR, but as you know i comment as soon as an idea come to me ;)
I think that adding 3 more special case to the already existing "mono" and "stereo" and handling the other in a more general way should be fine, suffisant.

Sami32 replied:
No worry, we have elephant memory 😉

Nadahar replied:
Don't include me, mine is more like that of a mouse.

Sami32 commented on src/main/java/net/pms/dlna/DLNAMediaInfo.java:
```diff
@@ -3404,4 +3408,38 @@ public static ScanOrder typeOf(String scanOrder) {
+			}
+
+			switch (value.trim().toLowerCase(Locale.ROOT)) {
+				case "c":
```
I guess i missed the meaning of such case

Nadahar replied:
@Sami32 It's *general*. The point is for it to be able to parse any "rate mode" from any source, so I just listed the values I could think of that anybody could use.

Sami32 replied:
ok, i never see this yet but who know ;)

Sami32 commented on src/main/java/net/pms/dlna/DLNAMediaInfo.java:
```diff
@@ -3404,4 +3408,38 @@ public static ScanOrder typeOf(String scanOrder) {
+				case "c":
+				case "cbr":
+				case "cfr":
+				case "const":
```
same here

Nadahar replied:
Same answer

Sami32 commented on src/main/java/net/pms/dlna/DLNAMediaInfo.java:
```diff
@@ -3404,4 +3408,38 @@ public static ScanOrder typeOf(String scanOrder) {
+				case "const":
+				case "constant":
+					return CONSTANT;
+				case "v":
```
same here

Sami32 commented on src/main/java/net/pms/dlna/LibMediaInfoParser.java:
```diff
@@ -932,65 +932,165 @@ public static String getAvcProfile(String value) {
+	public static int getSpecificID(String value) {
+		// If ID is given as 'streamID-substreamID' use the second (which is hopefully unique).
+		// For example in vob audio ID can be '189 (0xBD)-32 (0x80)' and text ID '189 (0xBD)-128 (0x20)'
+		int end = value.lastIndexOf("(0x");
```
This problematic handling way should enhanced later IMO

Nadahar replied:
You're probably right, it seems too simple. I don't know enough to know how the field is used to know what would be a proper implementation though.

Sami32 replied:
Hoping that could give you a brighter idea ;)

Nadahar replied:
@Sami32 I still don't quite understand this crazy logic. Why is this only used for OGG? Does this whole logic only apply to OGG?

I am also confused about the form - why does the current code include hexadecimal values while your example does not? Is it because of ```ID``` vs ```ID/String```?

If ```ID``` never return hexadecimal numbers, your code should work? So should the existing code it seems to me, although it's a bit crazy to go to all that trouble by using the ```ID/String``` value.

Sami32 replied:
Yes, this holy shit has been done only for OGG files, it is why i removed it and handled in a more general way, as it also made the EIA subtitle handling fail as your already saw the issue pending since months.

Sami32 replied:
The fact that you make me "return" to Java was not very easy for me, but as i finally feeled responsable for it ;P
You will find it in this commit, where you'll recognize some of your code ;)
https://github.com/UniversalMediaServer/UniversalMediaServer/commit/e1007ad1b52c639fb03e48281262a511f1d1a09c

Sami32 replied:
Yes, i agree, as i said it should be fixed later.

Sami32 replied:
Here the old link pointing at the database failure that such handling give:
https://github.com/UniversalMediaServer/UniversalMediaServer/issues/1232

Sami32 commented:
Sorry, i only read your message now.
No, as i met issue while trying to build it because the binaries site is down, but i finally take them from an other place but didn't get the time to test it, as i'm still attempting to review it first.
That look great 👍 
It seem even better than my code, but don't repeat it to anyone 😉

Sami32 commented on src/main/java/net/pms/dlna/LibMediaInfoParser.java:
```diff
@@ -200,11 +203,12 @@ public synchronized static void parse(DLNAMediaInfo media, InputFile inputFile,
-					currentAudioTrack.setSampleFrequency(getSampleFrequency(MI.Get(StreamType.Audio, i, "SamplingRate")));
-					currentAudioTrack.setBitRate(getBitrate(MI.Get(StreamType.Audio, i, "BitRate")));
+					currentAudioTrack.setAudioTrackTitleFromMetadata(MI.Get(StreamType.Audio, i, "Title").trim());
+					currentAudioTrack.setNumberOfChannels(parseNumberOfChannels(MI.Get(StreamType.Audio, i, "Channel(s) Original")));
```
Typo,  it should have been: `"Channel(s)_Original"`

Sami32 replied:
```
value = MI.Get(audio, i, "Channel(s)_Original");
if (isNotBlank(value)) {
	currentAudioTrack.getAudioProperties().setNumberOfChannels(MI.Get(audio, i, "Channel(s)_Original"));
} else {
	currentAudioTrack.getAudioProperties().setNumberOfChannels(MI.Get(audio, i, "Channel(s)"));
}
```

Nadahar replied:
@Sami32 I have some unpushed bugfixes, and that's one of them.

Sami32 replied:
Good 👍 
I just take some sample from my old unpublished commits when reviewing the code make me remembering something 😉

Sami32 commented:
I still didn't finished to review but i don't get your delay handling, with default and minimal value.
I don't remember having done such thing, because it doesn't make sens to me as tsMuxeR, and maybe MEncoder, use this value only if different from zero if i remember correctly.

Nadahar commented:
@Sami32 The point with the delay handling is as with everything else, you need to have a value that indicates "no value"/"no known value"/"unknown". This is what null usually means, but since this is an ```int``` null doesn't exist.

The "standard" null-indicator used for integers is -1. That works fine as long as the integer can only be zero or positive. Since delay can be both positive and negative, -1 doesn't work for it. That's why I chose ```Integer.MIN_VALUE``` (```0x80000000```) as the "null value" instead.  I doubt the delay will ever get big enough for that to be a problem (-2147483648), so it should work just fine for that.

The whole point of a "null value" is to have a value that will never be an actual value, so that you can tell it apart from a real value. -1 works fine for values that can only be positive for that reason.

Sami32 commented:
ok, but an "unknown" delay value shouldn't be logged at all IMO.

Nadahar commented:
@Sami32 I've pushed my fixes. I haven't looked into the ID as I still don't quite understand it. I wasn't going to do a rebase to make it simple, but then I found something I had forgotten to remove in a previous commit so I "had to". The "fixes" are still in a separate commit so they should be easy to review though.

Nadahar commented:
> Ok, but an "unknown" delay value shouldn't be logged at all IMO.

It shouldn't be either. I did some fixes around that in my "fixes". Are you sure it is still logged if unknown?

> As i'm at testing it, a delay of zero shouldn't be logged as the other delay related fields.

If you're talking about the parse logger, I haven't really changed anything related to that - but I can't see why ```0``` shouldn't be logged as it is a parsed value. It is not logged if MediaInfo doesn't return ```0```.

> FormatV and FormatP should be replaced by Profile and Version to be more meaningful IMO. Yes, i think that that PR has already been merged.

Now you're confusing me with thing that's unrelated. I'll see if I understand what you mean.

> It seem that Bitrate get double display with some codec like ALAC or "mp4A" codecs:

Yes, it was another bug in #20. I've fixed that in my "fixes".

Sami32 commented:
Ok, now that i finished testing for the other PR already merged you can detest me.

Sami32 commented:
```
DEBUG 2018-05-13 22:47:39.656 [Library Scanner] Error while getting cached information about MPEG2-AC3-2ch_OK.mpg, reparsing information: Column "BITRATEMODE" not found [42122-196]
TRACE 2018-05-13 22:47:39.656 [Library Scanner] 
org.h2.jdbc.JdbcSQLException: Column "BITRATEMODE" not found [42122-196]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.jdbc.JdbcResultSet.getColumnIndex(JdbcResultSet.java:3154)
	at org.h2.jdbc.JdbcResultSet.get(JdbcResultSet.java:3246)
	at org.h2.jdbc.JdbcResultSet.getObject(JdbcResultSet.java:533)
	at net.pms.dlna.DLNAMediaDatabase.getData(DLNAMediaDatabase.java:478)
	at net.pms.dlna.RealFile.resolve(RealFile.java:180)
	at net.pms.dlna.DLNAResource.syncResolve(DLNAResource.java:1527)
	at net.pms.dlna.DLNAResource.run(DLNAResource.java:1328)
	at net.pms.dlna.RealFile.isValid(RealFile.java:73)
	at net.pms.dlna.DLNAResource.addChild(DLNAResource.java:615)
	at net.pms.dlna.DLNAResource.addChild(DLNAResource.java:581)
	at net.pms.dlna.MapFile.manageFile(MapFile.java:281)
	at net.pms.dlna.MapFile.analyzeChildren(MapFile.java:333)
	at net.pms.dlna.RootFolder.scan(RootFolder.java:272)
	at net.pms.dlna.RootFolder.scan(RootFolder.java:282)
	at net.pms.dlna.RootFolder.scan(RootFolder.java:228)
	at net.pms.dlna.DLNAMediaDatabase.run(DLNAMediaDatabase.java:1028)
	at java.lang.Thread.run(Unknown Source)
```

Sami32 commented:
- The default values are not logged

Sami32 commented:
Because as we already agree, the default value can be omitted, a delay of 0 ms seem very useful information to have and what a surprise...

Nadahar commented:
@Sami32 Regarding the database error, try resetting your database. There are some issues with the way the "database version" is handled when switching between different branches that might have the same number, so I think it should be fixed if you simply reset it.

Nadahar commented:
> A delay of 0 ms seem very useful information to have and what a surprise... 
> Adding more informations like that should make the log and debugging too easy...

This looks like attempted sarcasm to me, but it's not me who put it there - it's you. If's from #20.

Sami32 commented:
I did reset it before, so i will try delete the database file.
BTW, shouldn't a Java automatic check for the database version be implemented, so that it could automatically reset if need or ask for?

Nadahar commented:
> BTW, shouldn't a Java automatic check for the database version be implemented, so that it could automatically reset if need or ask for?

That's already done. The problem is that it's too simplistic, so if one change to the database is made in branch A the version is bumped with 1, and if another change to the database is made in branch B the version is also bumped by 1. You then have two branches with the same version number but with incompatible databases.

Sami32 commented:
Ok.
That solved the database error issue.
Though still no default samplerate, bitdepth and bitrate values are logged as it should be when such values are `0`:
```
23:30:54 TRACE Matched format MPG to "F:\_DEVELOPPEMENT_TOOLS\!MEDIA_TEST_FILES\UMS Video Testing Suite (20150121)\TESTS\MPEG-PS\MPEG2_LPCM16_32_2.vob"
23:30:54 DEBUG Stopping process: ffmpeg.exe-180
23:30:55 DEBUG Stopping process: ffmpeg.exe-181
23:30:55 DEBUG Stopping process: ffmpeg.exe-182
23:30:55 DEBUG Stopping process: ffmpeg.exe-183
23:30:55 DEBUG Stopping process: ffmpeg.exe-184
23:30:55 DEBUG MediaInfo returned a blank value for the number of audio channels, returning -1
23:30:55 TRACE MediaInfo parsing results for "F:\_DEVELOPPEMENT_TOOLS\!MEDIA_TEST_FILES\UMS Video Testing Suite (20150121)\TESTS\MPEG-PS\MPEG2_LPCM16_32_2.vob":
  Format: "MPEG-PS"             Duration: 0:00:45.000         Overall Bitrate Mode: VBR     Overall Bitrate: 3190511
  Cover: No                     FPS: 25.000
  - Video track ID: 224
    Format: "MPEG Video"          FormatV: "Version 2"          FormatP: "Main@High"          Duration: 0:00:45.000
    BitRate Mode: VBR             Bitrate: 3127020              BitRate Maximum: 9800000      Width: 1920
    Height: 1080                  Colorimetry: 4:2:0            Chroma: 4:2:0                 PAR: 1.000
    DAR: 16:9                     Scan Type: Progressive        FPS: 25.000                   GOP: "N=15"
    Bit Depth: 8                  Delay: 540.000                Delay Source: Container       Delay Original: 0
    Delay O. Source: Stream
  - Audio track ID: 189-160
    Format: "PCM"                 Duration: 0:00:44.959         BitRate Mode: CBR             Bitrate: 0
    Channel(s): 2                 Samplerate: 0                 Bitrate: 0                    Bit Depth: 16
    Delay: 540.000                Delay Source: Container
23:30:55 TRACE Parsing results for file "MPEG2_LPCM16_32_2.vob": Container: MPEGPS, Size: 17946624, Video Bitrate: 3190511, Video Tracks: 1, Video Codec: mpeg2, Duration: 0:00:45.000, Video Resolution: 1920 x 1080, Display Aspect Ratio: 16:9, Scan Type: Progressive, Framerate: 25.000, Audio Tracks: 1 [Audio Codec: LPCM, Bitrate Mode: CONSTANT, Bits per Sample: 16, Delay: 0 ms], Mime Type: video/mpeg
```

Sami32 commented:
Ah, and just in case you've missed the issues added on the way in this post:
https://github.com/DigitalMediaServer/DigitalMediaServer/pull/43#issuecomment-388644859

Sami32 commented:
> If you're still talking about the parse logging, it behaves like you wanted it in #20. Not logging values that are equal to those chosen as defaults doesn't make a lot of sense to me, but as I said you could make the parse logger behave as you wanted.

No, i was meaning informations added in this PR are not logged like 23:30:55 DEBUG MediaInfo returned a blank value for the number of audio channels, returning -1
Nothing is said about the default values set.

Sami32 commented:
But yes, i didn't added them AFAICR, perhaps because i didn't thought about, or perhaps because i didn't knew how to do that.

Nadahar commented:
I just remembered that when I added bitrate mode for ```DLNAMediaAudio``` I forgot to add it to the database as well, this must also be done.

Sami32 commented:
A delay of 0 ms seem very useful information to have and what a surprise...
    Adding more informations like that should make the log and debugging too easy...

> This looks like attempted sarcasm to me, but it's not me who put it there - it's you. If's from #20.

It was ;)
Maybe, but did i said that i'm perfect? Perhaps it was my only one mistake ;P

Sami32 commented:
I didn't saw it, so 4 yes are always better, also if mine was late ;)

Nadahar commented:
>No, i was meaning informations added in this PR are not logged like 23:30:55 DEBUG MediaInfo returned a blank value for the number of audio channels, returning -1
>Nothing is said about the default values set in the debug log.

This should have been removed in my "fixes" commit. Are you still running the old version?

Sami32 commented:
Are you kidding? You know how i'm slow, so i only used the first build ;)
I didn't saw it in the fix commit i saw but i'll ve to try with the latest commit then.

Sami32 commented:
But as you said "removed" i guess that nothing as been debug added?

Sami32 commented:
What a pain to build on this branch...

Sami32 commented:
It look great, and i better say it is perfect or you will kick my ass 😉 

But i still think that adding something like:
TRACE A default bitrate value of 8 kb/s has been added because of a faulty (0) value detected.
TRACE A default sampling rate value of 8000 Hz has been added because of a faulty (0) value detected.
TRACE A default bit depth value of 16 has been added because of a faulty (0) value detected.

Sami32 commented:
Hum...it seem that duration parsing is not too resilient:
`00:28:08 WARN  Could not parse duration from "6687 / 6687"`

Nadahar commented:
@Sami32 Why would there be two different durations? How should we deal with that, which one should we use?

Sami32 commented:
`Duration                                 : 6 s 687 ms / 6 s 687 ms`

Just use one as the other is just the same. A MediaInfo bug i guess.
That mean that a duration parsing detecting `/` should be used IMO.

Nadahar commented:
@Sami32 I've added the forgotten database field and fixed the duration parser.

Nadahar commented:
@Sami32 If you confirm that the last commit works fine, I'll rebase the "fixes" into their proper places and merge.

Sami32 commented:
Ohoh...it seem that i've missed some of your comments yesterday.

Nadahar commented:
> Ohoh...it seem that i've missed some of your comments yesterday.

Tell me something new :stuck_out_tongue_closed_eyes:

Nadahar commented:
@Sami32 Regarding your logging of the use of default values, it isn't possible with the current logic. If the defaults were applied in the code that needed it, like the DIDL generation code, it should be logged by that code. That will have to be some future project though, perhaps when the DIDL generation is refactored.

Sami32 commented:
Oh i see why i get confused about the delay information displayed, as it is a "special" case i've missed that 2 values are returned and that we displayed only one. So it is fine.
```
Delay                                    : 2891321.250
Delay                                    : 48 min 11 s
Delay                                    : 48 min 11 s 321 ms
Delay                                    : 48 min 11 s
Delay                                    : 00:48:11.321
Delay_DropFrame                          : No
Delay, origin                            : Container
Delay, origin                            : Container
Delay_Original                           : 6631040
Delay_Original                           : 1 h 50 min
Delay_Original                           : 1 h 50 min 31 s 40 ms
Delay_Original                           : 1 h 50 min
Delay_Original                           : 01:50:31.040
Delay_Original_Settings                  : drop_frame_flag=0 / closed_gop=1 / broken_link=0
Delay_Original_DropFrame                 : No
Delay_Original_Source                    : Stream
Time code of first frame                 : 01:50:31:01
Time code source                         : Group of pictures header
```

Nadahar commented:
Good, because I am even more confused when you start complaining about the stuff from #20 :innocent:

Sami32 commented:
Ok, my bad, i didn't knew that the default values logging will need a special implementation.

Nadahar commented:
@Sami32 Do you have any remaining issues that is actually related to this PR, or are with good to go?

Sami32 commented:
After having fun building it ;P to test the latest commits, it seem that we are god ;)

Nadahar commented:
Reviewed 2 of 9 files at r1, 10 of 18 files at r4, 1 of 1 files at r5, 5 of 8 files at r6, 2 of 3 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/digitalmediaserver/digitalmediaserver/43)*
<!-- Sent from Reviewable.io -->

Nadahar commented:
OK

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/digitalmediaserver/digitalmediaserver/43#-:-LCUV9Oc7buh9JHTU8E0:b32w0x)*
<!-- Sent from Reviewable.io -->

Sami32 commented:
Damn it...something get wrong with the bitrate displayed.
`00:10:23 TRACE Parsing results for file "hd_dtsma_5.1_DTS.dts": Container: DTS, Size: 7978244, Bitrate: -1, Bitrate Mode: VARIABLE, Duration: null, Audio Tracks: 1 [Audio Codec: DTS, Bitrate: 1509000, Bitrate Mode: VARIABLE, Bits per Sample: 16, Channels: 6, Sample Frequency: 48000 Hz], Mime Type: audio/vnd.dts.hd`

Sami32 commented:
Oh yes i remember that one, but thought that you already get it fixed long ago.
```
General
Complete name                            : F:\_DEVELOPPEMENT_TOOLS\!MEDIA_TEST_FILES\AUDIO\hd_dtsma_5.1_DTS.dts
Format                                   : DTS
Format/Info                              : Digital Theater Systems
Format profile                           : MA / Core
File size                                : 7.61 MiB
Overall bit rate mode                    : Variable / Constant

Audio
Format                                   : DTS
Format/Info                              : Digital Theater Systems
Format profile                           : MA / Core
Bit rate mode                            : Variable / Constant
Bit rate                                 : Unknown / 1 509 kb/s
Channel(s)                               : 6 channels
Channel positions                        : Front: L C R, Side: L R, LFE
Sampling rate                            : 48.0 kHz
Frame rate                               : 93.750 FPS (512 SPF)
Bit depth                                : 16 bits
Compression mode                         : Lossless / Lossy
```

Sami32 commented:
Now i remember why i wanted the RAW bitrate value ;)

Sami32 commented:
Well, actually the bitrate is possible to determine with MediaInfo, if i remember correctly, but that will ask to set the parsing to be set to 100% of the media...so you can imagine of infinitely long that could quickly become...

Sami32 commented:
At least you have set it correctly in the formatted parsing 👍 
```
00:10:23 TRACE MediaInfo parsing results for "F:\_DEVELOPPEMENT_TOOLS\!MEDIA_TEST_FILES\AUDIO\hd_dtsma_5.1_DTS.dts":
  Format: "DTS"                 Overall Bitrate Mode: VBR / CBR                             Cover: No
  - Audio track ID: 
    Format: "DTS"                 Profile: "MA / Core"          BitRate Mode: VBR / CBR       Bitrate: Unknown / 1509000
    Channel(s): 6                 Samplerate: 48000             Bit Depth: 16
```

Sami32 commented:
Sure i can understand that for you a bitrate of -1 look fine but that is different for me.

Nadahar commented:
Argh, you just can't let it go. It's "correctly" logged the other place too, it's just that it's showing different information than what you want to see. The ParseLogger logs what MediaInfo has parsed, while the "parsing results" shows what values DMS has *after* parsing MediaInfo's (or any other parser's) output. Thus, both are correct :wink:

Nadahar commented:
It's because it's NOT a bitrate that is logged, it's the value of the "bitrate" field in that instance of DLNAMediaInfo. That's a big difference, and that value is ```-1```.

Sami32 commented:
I guess it will never be for me ;P

Nadahar commented:
Me too :wink:
